### PR TITLE
Add default date for student timestamp #4684

### LINKED ISSUE #4684: Add default date for student timestamp
Set  2011.01.01 as default for createdAt and updatedAt in the getter functions

### PR DESCRIPTION
**Pull request by jkt001**
Fixes #4684 

I have modified the tests as well to check whether students with no createdAt and updatedAt timestamp to return the default timestamp as well.

### CODE REVIEW

File: src/main/java/teammates/common/datatransfer/StudentAttributes.java [hunk 1]
```diff
@@ -364,10 +364,16 @@ public String getStudentStatus() {
...
+        if (createdAt == null) {
+            createdAt = new Date(Const.DEFAULT_STUDENT_CREATION_DATE);
+        }
         return createdAt;
     }
 
     public Date getUpdatedAt() {
+        if (updatedAt == null) {
+            updatedAt = new Date(Const.DEFAULT_STUDENT_CREATION_DATE);
+        }
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: Two issues here:

1. You can do `if createdAt == null ? ... : ...` to make it shorter.
2. Your code is doing "if the field is null, update it to the default value first, then get it", does it match what "getCreatedAt" "seems" to do? If you implement it like 1 you can avoid this problem also.

File: src/main/java/teammates/common/util/Const.java [hunk 2]
```diff
@@ -1193,6 +1193,9 @@ private void _______values_with_special_meanings________________________(){}
     public static final Date TIME_REPRESENTS_LATER;
     public static final Date TIME_REPRESENTS_NOW;
     
+    //represents time (in milliseconds) from epoch (1/1/1970) to default student creation date (1/1/2011) in UTC
+    public static final long DEFAULT_STUDENT_CREATION_DATE = 1293840060000L;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Const.java]: Even if explained, this number looks "magic". Consider parsing a date string to a Date object and then getting the long value from it.

File: src/main/java/teammates/storage/entity/Student.java [hunk 3]
```diff
@@ -125,7 +129,10 @@ public void setCreatedAt(Date created) {
     }
     
     public Date getUpdatedAt() {
-        return updatedAt;
+        if (updatedAt == null) {
+            setCreatedAt(new Date(Const.DEFAULT_STUDENT_CREATION_DATE));
+         }
+         return updatedAt;
```
Comment (wkurniawan07) [inline src/main/java/teammates/storage/entity/Student.java]: Same issue with the one in attributes.

File: src/main/java/teammates/storage/entity/Student.java [hunk 4]
```diff
@@ -125,7 +129,10 @@ public void setCreatedAt(Date created) {
     }
     
     public Date getUpdatedAt() {
-        return updatedAt;
+        if (updatedAt == null) {
+            setCreatedAt(new Date(Const.DEFAULT_STUDENT_CREATION_DATE));
```
Comment (LowWeiLin) [inline src/main/java/teammates/storage/entity/Student.java]: Why do you setCreatedAt when this function is for updatedAt?

**Comment by LowWeiLin** [comment]
The createdAt and updatedAt fields should not be modified, as per what @wkurniawan07 commented.
Simply return the default value.

**Comment by jkt001** [comment]
Done.

File: src/test/java/teammates/test/cases/storage/StudentsDbTest.java [hunk 5]
```diff
@@ -33,6 +35,38 @@ public static void setupClass() throws Exception {
...
+        student.setCreated_NonProduction(null);
+        student.setUpdatedAt_NonProduction(null);
+        
+        Date defaultStudentCreationTimeStamp = Const.TIME_REPRESENTS_DEFAULT_STUDENT_CREATION_TIMESTAMP;
+        
+        ______TS("success : created defaultTimeStamp");
+        
+        assertEquals(defaultStudentCreationTimeStamp, student.getCreatedAt());
+        
+        ______TS("success : update defaultTimeStamp");
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/storage/StudentsDbTest.java]: The test case description is misleading here. Write as "defaultTimeStamp for updatedAt date" instead. Similarly for the createdAt date test case.

File: src/main/java/teammates/common/util/Const.java [hunk 6]
```diff
@@ -1192,13 +1192,15 @@ private void _______values_with_special_meanings________________________(){}
     public static final Date TIME_REPRESENTS_NEVER;
     public static final Date TIME_REPRESENTS_LATER;
     public static final Date TIME_REPRESENTS_NOW;
-    
+    public static final Date TIME_REPRESENTS_DEFAULT_STUDENT_CREATION_TIMESTAMP;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Const.java]: As per what @LowWeiLin mentioned, rename the variable to `TIME_REPRESENTS_DEFAULT__TIMESTAMP`.

File: src/test/java/teammates/test/cases/storage/StudentsDbTest.java [hunk 7]
```diff
@@ -33,6 +35,38 @@ public static void setupClass() throws Exception {
     }
     
     @Test
+    public void testDefaultTimestamp() throws InvalidParametersException, EntityAlreadyExistsException, EntityDoesNotExistException {        
+        ______TS("success : created");
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/storage/StudentsDbTest.java]: I don't think you need a TS description here. Creating the student attribute is just a preparatory procedure.

**Comment by wkurniawan07** [comment]
Few minor improvements and it's good to go.

**Comment by jkt001** [comment]
@wkurniawan07 I have renamed the constant and updated the test headers as well. For your review, please.

**Comment by wkurniawan07** [comment]
Looks good :+1:

**Comment by thyageshm** [comment]
@wkurniawan07 I think you have to help @jkt001 to merge this one into master

**Comment by thyageshm** [comment]
@wkurniawan07, please merge this one soon

**Pull request merged**